feat: Agreement retirement

rafaelmag110 opened the pull request:
## WHAT

Introduces a feature that enables a participant to retire an active contract agreement.

## WHY

There are cases where the cancellation or update of a contract agreement might be valid, i.e. as CatenaX use cases evolve the respective frame contracts also do, and legally speaking the old contracts are no longer valid after the new version is released. There should be a way to technically retire the old agreements, effectively terminating any ongoing transfer process and disabling any new transfer request on that agreement.

## FURTHER NOTES

Only in memory storage is provided for the AgreementsRetirementStore. An SQL based implementation will be provided in a following PR.

This feature only works from a provider perspective since no "agreement retirement message" is sent to the consumer. This only enables the technical retirement of an agreement after the official retirement is legally agreed upon by both parties.

Implements DR introduced in https://github.com/eclipse-tractusx/tractusx-edc/pull/1604
Closes https://github.com/eclipse-tractusx/sig-release/issues/777

bmg13 commented on edc-extensions/agreements/retirement-evaluation-core/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/defaults/DefaultStoreProviderExtension.java:
```diff
@@ -0,0 +1,41 @@
+import org.eclipse.edc.spi.system.ServiceExtension;
+import org.eclipse.tractusx.edc.agreements.retirement.spi.store.AgreementsRetirementStore;
+
+@Extension("Provides a default Agreements")
```
maybe 
```suggestion
@Extension("Provides a default Agreements Store")
```

bmg13 commented on edc-extensions/agreements/retirement-evaluation-core/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/function/AgreementsRetirementFunction.java:
```diff
@@ -0,0 +1,65 @@
+        this.store = store;
+    }
+
+    @Override
+    public boolean evaluate(Permission rule, PolicyContext policyContext) {
+        var agreement = policyContext.getContextData(ContractAgreement.class);
+        if (agreement == null) {
+            return true;
+        }
+        var result = store.findRetiredAgreements(createFilterQueryByAgreementId(agreement.getId()));
+
+        if (!result.getContent().isEmpty()) {
+            policyContext.reportProblem(String.format("Contract Agreement with ID=%s has been retired", agreement.getId()));
+
+        }
+
+        return true;
+    }
```
maybe a single return would be easier to read. personal preference, so left as mere suggestion
```suggestion
    @Override
    public boolean evaluate(Permission rule, PolicyContext policyContext) {
        var agreement = policyContext.getContextData(ContractAgreement.class);
        if (agreement != null) {
        
        var result = store.findRetiredAgreements(createFilterQueryByAgreementId(agreement.getId()));

        if (!result.getContent().isEmpty()) {
            policyContext.reportProblem(String.format("Contract Agreement with ID=%s has been retired", agreement.getId()));

        }
        }

        return true;
    }
```

rafaelmag110 replied:
[106ae62](https://github.com/eclipse-tractusx/tractusx-edc/pull/1615/commits/106ae6229b0bec53515d5fdb9b194795a925e5ae)

bmg13 commented on edc-extensions/agreements/retirement-evaluation-spi/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/spi/store/AgreementsRetirementStore.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.List;
+
+@ExtensionPoint
```
maybe add docs.

rafaelmag110 replied:
[8d0b9ec](https://github.com/eclipse-tractusx/tractusx-edc/pull/1615/commits/8d0b9ec726f4fde9a96445144995ff75f77fc2c1)

bmg13 commented on edc-extensions/agreements/retirement-evaluation-core/src/testFixtures/java/org/eclipse/tractusx/edc/agreements/retirement/store/AgreementsRetirementStoreTestBase.java:
```diff
@@ -0,0 +1,106 @@
+        var entry = createRetiredAgreementEntry(agreementId, "mock-reason");
+        store.save(entry);
+        var delete = store.delete(agreementId);
+        assertThat(delete.succeeded()).withFailMessage(delete::getFailureDetail).isTrue();
```
the ```getFailureDetail```should be empty, right?

rafaelmag110 replied:
I refactored the test because the `withFailMessage()` does not have the effect I expected it to have. Thanks for pointing this out.
[70c1003](https://github.com/eclipse-tractusx/tractusx-edc/pull/1615/commits/70c10038ac71342bf0d97d525b0962a0f0a2be6f)

bmg13 reviewed (commented):
just have some nits so far, all looks great

bmg13 commented on edc-extensions/agreements/retirement-evaluation-spi/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/spi/types/AgreementsRetirementEntry.java:
```diff
@@ -0,0 +1,102 @@
+            requireNonNull(entity.reason, AR_ENTRY_REASON);
+
+            if (entity.agreementRetirementDate == null) {
+                entity.agreementRetirementDate = Instant.now().toString();
```
shouldn't we already set here as a timestamp in accordance with the decision record?

ndr-brt replied:
Agree, this should be either a `long` or an `Instant` object.

rafaelmag110 replied:
You are right, thanks for making a point out of it. It's changed in: [46ef1b7](https://github.com/eclipse-tractusx/tractusx-edc/pull/1615/commits/46ef1b7c404e70567478a983fe14a6f11c7208f8)

ndr-brt commented on edc-extensions/agreements/retirement-evaluation-api/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/api/v3/AgreementsRetirementApiV3Controller.java:
```diff
@@ -0,0 +1,113 @@
+
+@Consumes(APPLICATION_JSON)
+@Produces(APPLICATION_JSON)
+@Path("/v3.1alpha/retireagreements")
```
I know that this is subtle and the convention is still not very well established, but there are already some new features in 3.1alpha, I was wondering if this should be filed under `v3.2alpha `.
Maybe we can leave here for now, but's something to discuss further.

rafaelmag110 replied:
Do we leave it in 3.1 or shift to 3.2?

ndr-brt replied:
we can just leave 3.1-alpha, as there's no timeline to stabilize it anyway.

ndr-brt commented on edc-extensions/agreements/retirement-evaluation-api/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/api/v3/AgreementsRetirementApiV3Controller.java:
```diff
@@ -0,0 +1,113 @@
+@Path("/v3.1alpha/retireagreements")
+public class AgreementsRetirementApiV3Controller implements AgreementsRetirementApiV3 {
+
+    private final AgreementsRetirementStore service;
```
```suggestion
    private final AgreementsRetirementStore store;
```

we usually put a `service` in between the store and the controller.
in this case it could look as unnecessary, but it has multiple purposes: 
- speak domain language (in this case will be `retire` and `reactivate` instead of `save` and `delete`)
- set the transaction context
- run domain validations if needed (e.g., return ServiceResult.NOT_FOUND when somebody tries to reactivate.)

bmg13 commented on settings.gradle.kts:
```diff
@@ -65,6 +69,7 @@ include(":edc-tests:edc-controlplane:catalog-tests")
 include(":edc-tests:edc-controlplane:transfer-tests")
 include(":edc-tests:edc-controlplane:iatp-tests")
 include(":edc-tests:edc-controlplane:policy-tests")
+include("edc-tests:edc-controlplane:agreement-retirement-tests")
```
```suggestion
include(":edc-tests:edc-controlplane:agreement-retirement-tests")
```

ndr-brt commented on edc-extensions/agreements/retirement-evaluation-core/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/AgreementsRetirementFunctionExtension.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    public void initialize(ServiceExtensionContext context) {
+        var function = new AgreementsRetirementFunction(store);
+        policyEngine.registerFunction(TRANSFER_SCOPE, Permission.class, function);
+        policyEngine.registerFunction(POLICY_MONITOR_SCOPE, Permission.class, function);
```
This works only because the policy we used in the test has a `Permission`, but if you submit a policy with a `Prohibition` instead it would not pass through the function.

Maybe the function would work better implemented as a "preValidator" (if it is retired, it won't bother running any function, just make the evaluation fail.

ndr-brt commented on edc-extensions/agreements/retirement-evaluation-core/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/function/AgreementsRetirementFunction.java:
```diff
@@ -0,0 +1,65 @@
+    public boolean evaluate(Permission rule, PolicyContext policyContext) {
+        var agreement = policyContext.getContextData(ContractAgreement.class);
+        if (agreement == null) {
+            return true;
```
these `return true` do definitely ring a bell.

ndr-brt commented on edc-extensions/agreements/retirement-evaluation-core/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/function/AgreementsRetirementFunction.java:
```diff
@@ -0,0 +1,65 @@
+        if (agreement == null) {
+            return true;
+        }
+        var result = store.findRetiredAgreements(createFilterQueryByAgreementId(agreement.getId()));
```
here's the service could be useful to abstract things a little and offer a 
```ServiceResult<boolean> isRetired(String agreementId)``` method.

rafaelmag110 commented on edc-extensions/agreements/retirement-evaluation-api/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/api/transform/JsonObjectFromAgreementRetirementTransformer.java:
```diff
@@ -0,0 +1,49 @@
+
+public class JsonObjectFromAgreementRetirementTransformer extends AbstractJsonLdTransformer<AgreementsRetirementEntry, JsonObject> {
+
+    JsonBuilderFactory jsonFactory;
```
[106ae62](https://github.com/eclipse-tractusx/tractusx-edc/pull/1615/commits/106ae6229b0bec53515d5fdb9b194795a925e5ae)

rafaelmag110 commented:
@ndr-brt I took your advise and introduced a service in between. I felt it was not needed before, but indeed it makes the API controller code more clearer.
I also took your advise and changed the policy function to a preValidator instead. The validator method is part of the new service. As such we don't need the policy function anymore.

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-spi/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/spi/service/AgreementsRetirementService.java:
```diff
@@ -0,0 +1,65 @@
+     * @param context a valid {@link PolicyContext}
+     * @return true if it exists, false otherwise.
+     */
+    boolean isRetired(PolicyContext context);
```
I would not use the `PolicyContext` here but `ContractAgreement` directly

wolf4ood commented on edc-tests/edc-controlplane/agreement-retirement-tests/src/test/java/org/eclipse/tractusx/edc/tests/agreement/retirement/RetireAgreementTest.java:
```diff
@@ -0,0 +1,168 @@
+
+        @Test
+        @DisplayName("Verify all existing TPs related to an agreement are terminated upon its retirement")
+        void retireAgreement_shouldCloseTransferProcesses() {
```
Since we have a `reactivate` option i would also tests that we can start TP after contract reactivation

rafaelmag110 replied:
that is tested in the same test, at the end of it

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-api/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/api/transform/JsonObjectFromAgreementRetirementTransformer.java:
```diff
@@ -0,0 +1,49 @@
+    @Override
+    public @Nullable JsonObject transform(@NotNull AgreementsRetirementEntry entry, @NotNull TransformerContext transformerContext) {
+        return jsonFactory.createObjectBuilder()
+                .add(AgreementsRetirementEntry.AR_ENTRY_AGREEMENT_ID, entry.getAgreementId())
+                .add(AgreementsRetirementEntry.AR_ENTRY_REASON, entry.getReason())
+                .add(AgreementsRetirementEntry.AR_ENTRY_RETIREMENT_DATE, entry.getAgreementRetirementDate())
```
static import

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-api/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/api/transform/JsonObjectToAgreementsRetirementEntryTransformer.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public @Nullable AgreementsRetirementEntry transform(@NotNull JsonObject jsonObject, @NotNull TransformerContext context) {
+        var entryBuilder = AgreementsRetirementEntry.Builder.newInstance();
+        entryBuilder.withAgreementId(transformString(jsonObject.get(AgreementsRetirementEntry.AR_ENTRY_AGREEMENT_ID), context));
+        entryBuilder.withReason(transformString(jsonObject.get(AgreementsRetirementEntry.AR_ENTRY_REASON), context));
```
static import

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-api/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/api/v3/AgreementsRetirementApiV3Controller.java:
```diff
@@ -0,0 +1,109 @@
+    @POST
+    @Override
+    public void retireAgreementV3(@RequestBody JsonObject entry) {
+        validator.validate(AgreementsRetirementEntry.AR_ENTRY_TYPE, entry).orElseThrow(ValidationFailureException::new);
```
static import

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-api/src/test/java/org/eclipse/tractusx/edc/agreements/retirement/api/transform/JsonObjectFromAgreementRetirementTransformerTest.java:
```diff
@@ -0,0 +1,61 @@
+        var result = transformer.transform(entry, context);
+
+        assertThat(result).isNotNull();
+        assertThat(result.getString(AgreementsRetirementEntry.AR_ENTRY_AGREEMENT_ID)).isEqualTo("agreementId");
+        assertThat(result.getString(AgreementsRetirementEntry.AR_ENTRY_REASON)).isEqualTo("long-reason");
+        assertThat(result.getJsonNumber(AgreementsRetirementEntry.AR_ENTRY_RETIREMENT_DATE)).isNotNull();
```
static import

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-api/src/test/java/org/eclipse/tractusx/edc/agreements/retirement/api/v3/AgreementsRetirementApiV3ControllerTest.java:
```diff
@@ -0,0 +1,222 @@
+
+    private JsonObject createJsonRetirementEntry(String agreementId) {
+        return Json.createObjectBuilder()
+                .add(TYPE, AgreementsRetirementEntry.AR_ENTRY_TYPE)
+                .add(AgreementsRetirementEntry.AR_ENTRY_AGREEMENT_ID, agreementId)
+                .add(AgreementsRetirementEntry.AR_ENTRY_REASON, "long-reason")
+                .add(AgreementsRetirementEntry.AR_ENTRY_RETIREMENT_DATE, Instant.now().toString())
```
static import

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-api/src/test/java/org/eclipse/tractusx/edc/agreements/retirement/api/v3/AgreementsRetirementApiV3ControllerTest.java:
```diff
@@ -0,0 +1,222 @@
+
+    private JsonObject createFilterQuerySpecJson(String agreement1) {
+        return Json.createObjectBuilder()
+                .add(TYPE, QuerySpec.EDC_QUERY_SPEC_TYPE)
+                .add(QuerySpec.EDC_QUERY_SPEC_FILTER_EXPRESSION,
+                        Json.createObjectBuilder()
+                                .add(Criterion.CRITERION_OPERAND_LEFT, AgreementsRetirementEntry.AR_ENTRY_AGREEMENT_ID)
+                                .add(Criterion.CRITERION_OPERATOR, "=")
+                                .add(Criterion.CRITERION_OPERAND_RIGHT, agreement1)
+                                .build())
```
static import

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-core/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/defaults/DefaultAgreementsRetirementService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    @Override
+    public boolean isRetired(PolicyContext context) {
```
as stated in the interface i would use `ContractAgreement` and let the caller handle the failure within the policy evaluation

wolf4ood reviewed (changes requested):
Smol nits, but i would use `ContractAgreement` in the service instead of the `PolicyContext`

rafaelmag110 commented:
@ndr-brt I address your suggestions. The `AgreementsRetirementService` was moved into a separate extension is not provided as default.
I also refactored the asserts on a `AbstractResult` to use the class from `edc junit`. Please have a look.

wolf4ood commented on edc-extensions/agreements/retirement-evaluation-spi/src/main/java/org/eclipse/tractusx/edc/agreements/retirement/spi/service/AgreementsRetirementService.java:
```diff
@@ -0,0 +1,65 @@
+public interface AgreementsRetirementService {
+
+    /**
+     * Within a given {@link PolicyContext}, verifies if the attached contract agreement exists in {@link org.eclipse.tractusx.edc.agreements.retirement.spi.store.AgreementsRetirementStore}.
```
leftover from refactoring

ndr-brt reviewed (approved):
:rocket:

wolf4ood commented:
@rafaelmag110 just regenerate the DEPENDENCIES file then we can merge it :)

rafaelmag110 commented:
@wolf4ood 
all ✅ 
🚀
